lib/render/mayaToHydra/renderOverride.cpp

```diff
@@ -0,0 +1,842 @@
+    TF_DEBUG(HDMAYA_RENDEROVERRIDE_RESOURCES)
+        .Msg(
+            "MtohRenderOverride created (%s - %s - %s)\n",
+            _rendererDesc.rendererName.GetText(),
+            _rendererDesc.overrideName.GetText(),
+            _rendererDesc.displayName.GetText());
+    _needsClear.store(false);
+    HdMayaDelegateRegistry::InstallDelegatesChangedSignal(
+        [this]() { _needsClear.store(true); });
+    _ID = SdfPath("/HdMayaViewportRenderer")
+              .AppendChild(TfToken(TfStringPrintf(
+                  "_HdMaya_%s_%p", desc.rendererName.GetText(), this)));
+
+    MStatus status;
+    auto id = MSceneMessage::addCallback(
+        MSceneMessage::kBeforeNew, _ClearHydraCallback, this, &status);
+    if (status) { _callbacks.push_back(id); }
+    id = MSceneMessage::addCallback(
+        MSceneMessage::kBeforeOpen, _ClearHydraCallback, this, &status);
+    if (status) { _callbacks.push_back(id); }
+    id = MEventMessage::addEventCallback(
+        MString("SelectionChanged"), _SelectionChangedCallback, this, &status);
+    if (status) { _callbacks.push_back(id); }
+
+    id = MTimerMessage::addTimerCallback(
+        1.0f / 10.0f, _TimerCallback, this, &status);
+    if (status) { _callbacks.push_back(id); }
+
+    _defaultLight.SetSpecular(GfVec4f(0.0f));
+    _defaultLight.SetAmbient(GfVec4f(0.0f));
+
+    {
+        std::lock_guard<std::mutex> lock(_allInstancesMutex);
+        _allInstances.push_back(this);
+    }
+
+    _globals = MtohGetRenderGlobals();
+
+#if WANT_UFE_BUILD
+    const UFE_NS::GlobalSelection::Ptr& ufeSelection =
+        UFE_NS::GlobalSelection::get();
+    if (ufeSelection) {
+        _ufeSelectionObserver = std::make_shared<UfeSelectionObserver>(*this);
+        ufeSelection->addObserver(_ufeSelectionObserver);
+    }
+#endif // WANT_UFE_BUILD
+}
+
+MtohRenderOverride::~MtohRenderOverride() {
+    TF_DEBUG(HDMAYA_RENDEROVERRIDE_RESOURCES)
+        .Msg(
+            "MtohRenderOverride destroyed (%s - %s - %s)\n",
+            _rendererDesc.rendererName.GetText(),
+            _rendererDesc.overrideName.GetText(),
+            _rendererDesc.displayName.GetText());
+
+    ClearHydraResources();
+
+    for (auto operation : _operations) { delete operation; }
+
+    for (auto callback : _callbacks) { MMessage::removeCallback(callback); }
+    for (auto& panelAndCallbacks : _renderPanelCallbacks) {
+        MMessage::removeCallbacks(panelAndCallbacks.second);
+    }
+
+    {
+        std::lock_guard<std::mutex> lock(_allInstancesMutex);
+        _allInstances.erase(
+            std::remove(_allInstances.begin(), _allInstances.end(), this),
+            _allInstances.end());
+    }
+
+#if WANT_UFE_BUILD
+    const UFE_NS::GlobalSelection::Ptr& ufeSelection =
+        UFE_NS::GlobalSelection::get();
+    if (ufeSelection) {
+        ufeSelection->removeObserver(_ufeSelectionObserver);
+        _ufeSelectionObserver = nullptr;
+    }
+#endif // WANT_UFE_BUILD
```

Comment:
@ppt-adsk Last time I heard from you that UFE can automatically remove dead observers?

Reply:
Yes, on next notification.  Observers are kept by weak pointer, and on next notification, the weak pointer lock will fail, and the observer will be removed.  This should be made clearer in the UFE docs.

Reply:
So, I take it then that you think we should remove the `removeObserver` call, then? I guess we could potentially remove this entire WANT_UFE_BUILD block, then?

Is there a downside to removing it ourselves? Is it less performant?

Reply:
There is no downside, but if you remove it the code can look cleaner.
 
WANT_UFE_BUILD won’t be removed because Maya 2018 is also in our list to support maya-usd and it doesn’t ship with UFE.

Reply:
I just meant that, if we're not removing the observer, the only thing left that we're doing is setting `_ufeSelectionObserver = nullptr` - which isn't really necessary, since this is the destructor.  So we can remove the whole thing.

Did that here:

https://github.com/Autodesk/maya-usd/pull/231/commits/d31f4e389f91e0251d9895cc2c9ad2016b37d9c3